graphblas/core/matrix.py

```diff
@@ -1335,30 +1336,131 @@ def from_dcsc(
+                )
+            if backend == "suitesparse":
+                # Should we try to handle F-contiguous data w/o a copy?
+                return cls.ss.import_fullr(
+                    values, dtype=dtype, nrows=nrows, ncols=ncols, is_iso=True, name=name
+                )
+            rv = cls(dtype, nrows=nrows, ncols=ncols, name=name)
+            rv << values
+            return rv
+        if values.ndim == 1:
+            raise ValueError("A 2d array or scalar is required to create a dense Matrix")
+        if values.ndim == 2 and dtype.np_type.subdtype is not None:
             raise ValueError("A >2d array is required to create a dense Matrix with subdtype")
-        nrows, ncols, *rest = values.shape
-        cols, rows = np.meshgrid(
-            np.arange(ncols, dtype=np.uint64),
-            np.arange(nrows, dtype=np.uint64),
-        )
-        rows = rows.ravel()
-        cols = cols.ravel()
-        if values.ndim > 2:
-            values = values.reshape([nrows * ncols, *rest])
+        if values.ndim > 2 and dtype.np_type.subdtype is None:
+            raise ValueError(f"values array must be 2d to create dense Matrix with dtype {dtype}")
+        if backend == "suitesparse":
+            rv = cls.ss.import_fullr(values, dtype=dtype, name=name)
+            nrows2, ncols2 = rv.shape
         else:
-            values = values.ravel()
-        return cls.from_coo(rows, cols, values, dtype, nrows=nrows, ncols=ncols, name=name)
+            nrows2, ncols2, *rest = values.shape
+            indptr = np.arange(0, nrows2 * ncols2 + 1, ncols2, dtype=np.uint64)
+            cols = np.repeat(np.arange(ncols2, dtype=np.uint64)[None, :], nrows2, 0).ravel()
+            if rest:  # sub-array dtype
+                values = values.reshape(nrows2 * ncols2, *rest)
+            else:
+                values = values.ravel()
+            rv = cls.from_csr(
+                indptr,
+                cols,
+                values,
+                dtype,
+                ncols=ncols2,
+                name=name,
+            )
+        if nrows is not None and nrows != nrows2 or ncols is not None and ncols != ncols2:
+            rv.resize(nrows2 if nrows is None else nrows, ncols2 if ncols is None else ncols)
+        return rv
+
+    def to_dense(self, fill_value=None, dtype=None):
+        """Convert Matrix to NumPy array of the same shape with missing values filled.
+
+        Parameters
+        ----------
+        fill_value : scalar, optional
+            Value used to fill missing values. This is required if there are missing values.
+        dtype :
+            Requested dtype for the output values array.
+
+        See Also
+        --------
+        to_coo
+        to_dicts
+        to_edgelist
+        from_dense
+        io.to_numpy
+
+        Returns
+        -------
+        np.ndarray
+        """
+        max_nvals = self._nrows * self._ncols
```

Comment:
Not sure if this is feasible or not, but we could add a check for very large matrix dimensions: if the dimensions of the matrix are such that their dense representation would exceed the available memory, then it might be worth warning the user and raising an error.

I remember `dask.array` had a similar warning when computing large arrays.

Reply:
Actually it turns out it was from `numpy` rather than `dask`: https://github.com/numpy/numpy/blob/8ace612c3b1bbf0bd23f10556492908afdc97207/numpy/core/_exceptions.py#L117

Reply:
Very good point. I added a warning in the docstring of `to_dense`, and I changed the implementation so it will quickly raise `OutOfMemory` when using SuiteSparse:GraphBLAS. Is this enough?

How would we choose the threshold for when to raise a warning for large output arrays? Do we just choose a number such as N GB, or base it on total memory or available memory?

Reply:
If I understand correctly, warning in the doc string is sufficient, because numpy will raise its own MemoryError if there is not enough available memory.